gorilla/server.go

```diff
@@ -5,113 +5,45 @@ package gorilla
-	}
-}
-
-// MWSpanObserver returns a MWOption that observe the span
-// for the server-side span.
-func MWSpanObserver(f func(span opentracing.Span, r *http.Request)) MWOption {
-	return func(options *mwOptions) {
-		options.spanObserver = f
-	}
-}
+const defaultComponentName = "github.com/gorilla/mux"
 
 // Middleware wraps an http.Handler and traces incoming requests.
 // Additionally, it adds the span to the request's context.
 //
 // By default, the operation name of the spans is set to "HTTP {method}".
 // This can be overriden with options.
 //
-// Example:
-// 	 http.ListenAndServe("localhost:80", gorilla.Middleware(tracer, http.DefaultServeMux))
-//
 // The options allow fine tuning the behavior of the middleware.
 //
 // Example:
+//   pattern := "/api/custerms/{id}"
 //   mw := gorilla.Middleware(
 //      tracer,
-//      http.DefaultServeMux,
-//      gorilla.OperationNameFunc(func(r *http.Request) string {
-//	        return "HTTP " + r.Method + ":/api/customers"
+//      nethttp.OperationNameFunc(func(r *http.Request) string {
+//	        return r.Proto + " " + r.Method + ":" + pattern
 //      }),
-//      gorilla.MWSpanObserver(func(sp opentracing.Span, r *http.Request) {
+//      nethttp.MWSpanObserver(func(sp opentracing.Span, r *http.Request) {
 //			sp.SetTag("http.uri", r.URL.EscapedPath())
 //		}),
 //   )
-func Middleware(tr opentracing.Tracer, h http.Handler, options ...MWOption) http.Handler {
-	opts := mwOptions{
-		opNameFunc: func(r *http.Request) string {
-			return "HTTP " + r.Method
-		},
-		spanObserver: func(span opentracing.Span, r *http.Request) {},
-	}
-	for _, opt := range options {
-		opt(&opts)
-	}
-	fn := func(w http.ResponseWriter, r *http.Request) {
-		ctx, _ := tr.Extract(opentracing.HTTPHeaders, opentracing.HTTPHeadersCarrier(r.Header))
-		sp := tr.StartSpan(opts.opNameFunc(r), ext.RPCServerOption(ctx))
-		ext.HTTPMethod.Set(sp, r.Method)
-		ext.HTTPUrl.Set(sp, r.URL.String())
-		opts.spanObserver(sp, r)
-
-		// set component name, use "net/http" if caller does not specify
-		componentName := opts.componentName
-		if componentName == "" {
-			componentName = defaultComponentName
-		}
-		ext.Component.Set(sp, componentName)
-
-		w = &statusCodeTracker{w, 200}
-		gorillaCtx := muxcontext.GetAll(r)
-		r = r.WithContext(opentracing.ContextWithSpan(r.Context(), sp))
-
-		//set mux context data
-		for key, value := range gorillaCtx {
-			muxcontext.Set(r,key,value)
+//   r := mux.NewRouter()
+//   r.HandleFunc(pattern, handler)
+//   r.Use(middleware.With)
+func Middleware(tr opentracing.Tracer, h http.Handler, options ...nethttp.MWOption) http.Handler {
+	opNameFunc := func(r *http.Request) string {
+		if tpl, err := mux.CurrentRoute(r).GetPathTemplate(); nil == err {
+			return r.Proto + " " + r.Method + " " + tpl
 		}
-
-		h.ServeHTTP(w, r)
-
-		ext.HTTPStatusCode.Set(sp, uint16(w.(*statusCodeTracker).status))
-		sp.Finish()
+		return r.Proto + " " + r.Method + " " + r.URL.Path 
```

Comment:
using URL as operation name is never a good idea, consider `/api/user/234234-123384-234234ef`

Reply:
I would change it to `return r.Proto + " " + r.Method`. 

Or it would be even better if nethttp allowed this function to return empty string and then the default would be whatever nethttp normally does.

Reply:
In this case, like /api/user/234234-123384-234234ef,  I thought  it probably matches  ‘r.Proto + " " + r.Method + " " + path template’.